rust/routee-compass-powertrain/src/model/energy_traversal_model.rs

```diff
@@ -60,18 +61,31 @@ impl TraversalModel for EnergyTraversalModel {
         // perform vehicle energy traversal
         let grade = get_grade(&self.energy_model_service.grade_table, edge.edge_id)?;
 
-        let distance_in_time_model_unit = BASE_DISTANCE_UNIT.convert(
-            &edge.distance,
+        let mut distance_in_energy_model_unit = Cow::Borrowed(&edge.distance);
+        baseunit::DISTANCE_UNIT.convert(
+            &mut distance_in_energy_model_unit,
             &self
                 .energy_model_service
                 .time_model_speed_unit
                 .associated_distance_unit(),
-        );
-        let speed = Speed::from((distance_in_time_model_unit, time_delta));
+        )?;
+        let speed_tuple = Speed::from_distance_and_time(
+            (
+                &distance_in_energy_model_unit,
+                &self.energy_model_service.distance_unit,
```

Comment:
I think I found the issue. When creating the speed here, we're telling it it's in the `energy_model_service.distance` unit (miles) but that distance has actually been converted to `energy_model_service.time_model_speed_unit.associated_distance_unit()` (kilometers). If I change like 75 to `&self.energy_model_service.time_model_speed_unit.associated_distance_unit(),`, the outputs are matching what we see in the main branch.

I think this highlights one of the main issues with the unit system that we should address which is that it's easy to mix up units if they're not explicitly paired together. I wonder if we can put together an abstraction such that we're required to pass around units with values.

Reply:
Nice catch. I agree if we could do that things would make more sense, but then we may greatly increase allocations by shipping around unit enum instances with every arithmetic operation. The current design is trying to avoid all of that cloning. I would love to see other ideas.

Reply:
Yeah that's a good point. I haven't spent enough time thinking about this to propose a good solution but hopefully we can eventually work towards something like this.